### PR TITLE
kyma-operator uses Helm3

### PR DESCRIPTION
**Pull request by strekm**
**Description**

Changes proposed in this pull request:

- bumped helm to version 3.2.1 in kyma-operator

**Related issue(s)**
see #8517

### CODE REVIEW

File: components/kyma-operator/pkg/kymahelm/wait.go [hunk 1]
```diff
@@ -73,53 +73,57 @@ func (hc *Client) WaitForCondition(releaseName string, pf WaitPredicateFunc, opt
...
 			}
 			log.Printf("Error while waiting for release delete: %s", getStatusRespErr.Error())
 			//Continue waiting
 			return false, nil
 		}
 
-		if resp != nil {
-			log.Printf("Waiting for release delete: release status: %s/%s: %s", resp.Namespace, resp.Name, resp.Info.Status.Code.String())
+		if relStatus == nil {
+			return false, errors.New("release status in nil")
```
Comment (Demonsthere) [inline components/kyma-operator/pkg/kymahelm/wait.go:92]:
```suggestion
			return false, errors.New("release status is nil")
```

File: components/kyma-operator/pkg/kymahelm/wait.go [hunk 2]
```diff
@@ -73,53 +73,57 @@ func (hc *Client) WaitForCondition(releaseName string, pf WaitPredicateFunc, opt
...
 			return false, nil
 		}
 
-		if resp != nil {
-			if resp.Info.Status.Code.String() == "DEPLOYED" {
-				return true, nil
-			}
-			log.Printf("Waiting for release rollback: release status: %s/%s: %s", resp.Namespace, resp.Name, resp.Info.Status.Code.String())
+		if relStatus == nil {
+			return false, errors.New("release status in nil")
```
Comment (Demonsthere) [inline components/kyma-operator/pkg/kymahelm/wait.go:114]:
```suggestion
			return false, errors.New("release status is nil")
```

File: components/kyma-operator/pkg/kymaoperation/steps/upgrade.go [hunk 3]
```diff
@@ -4,11 +4,13 @@ import (
...
 	"fmt"
 	"log"
+
+	"github.com/kyma-project/kyma/components/kyma-operator/pkg/kymahelm"
 )
 
 type upgradeStep struct {
 	installStep
-	deployedRevision    int32
+	// deployedRevision    int //todo: deployed revision is refreshed before rollback, remove this field.
```
Comment (Demonsthere) [inline components/kyma-operator/pkg/kymaoperation/steps/upgrade.go:13]:
```suggestion
```

File: components/kyma-operator/pkg/kymaoperation/steps/factory.go [hunk 4]
```diff
@@ -84,8 +84,8 @@ func (isf installStepFactory) newStep(component v1alpha1.KymaComponent) (Step, e
 
 		if isUpgrade {
 			return upgradeStep{
-				installStep:         inststp,
-				deployedRevision:    relStatus.LastDeployedRevision,
+				installStep: inststp,
+				// deployedRevision:    relStatus.LastDeployedRevision, todo: deployed revision is refreshed before rollback, no need to set it here.
```
Comment (Demonsthere) [inline components/kyma-operator/pkg/kymaoperation/steps/factory.go:88]:
```suggestion
```

File: components/kyma-operator/cmd/operator/main.go [hunk 5]
```diff
@@ -53,16 +53,18 @@ func main() {
...
 	kubeconfig := flag.String("kubeconfig", "", "Path to a kubeconfig file")
-	helmHost := flag.String("helmhost", "tiller-deploy.kube-system.svc.cluster.local:44134", "Helm host")
+	//helmHost := flag.String("helmhost", "tiller-deploy.kube-system.svc.cluster.local:44134", "Helm host")
 	kymaDir := flag.String("kymadir", "/kyma", "Directory where kyma packages will be extracted")
-	tlsKey := flag.String("tillerTLSKey", "/etc/certs/tls.key", "Path to TLS key file")
-	tlsCrt := flag.String("tillerTLSCrt", "/etc/certs/tls.crt", "Path to TLS cert file")
-	TLSInsecureSkipVerify := flag.Bool("tillerTLSInsecureSkipVerify", false, "Disable verification of Tiller TLS cert")
+	//tlsKey := flag.String("tillerTLSKey", "/etc/certs/tls.key", "Path to TLS key file")
+	//tlsCrt := flag.String("tillerTLSCrt", "/etc/certs/tls.crt", "Path to TLS cert file")
+	//TLSInsecureSkipVerify := flag.Bool("tillerTLSInsecureSkipVerify", false, "Disable verification of Tiller TLS cert")
```
Comment (Demonsthere) [inline components/kyma-operator/cmd/operator/main.go:60]:
```suggestion
```

File: components/kyma-operator/cmd/operator/main.go [hunk 6]
```diff
@@ -53,16 +53,18 @@ func main() {
 	env.InitConfig()
 
 	kubeconfig := flag.String("kubeconfig", "", "Path to a kubeconfig file")
-	helmHost := flag.String("helmhost", "tiller-deploy.kube-system.svc.cluster.local:44134", "Helm host")
+	//helmHost := flag.String("helmhost", "tiller-deploy.kube-system.svc.cluster.local:44134", "Helm host")
```
Comment (Demonsthere) [inline components/kyma-operator/cmd/operator/main.go:56]:
```suggestion
```

File: components/kyma-operator/pkg/kymaoperation/executor.go [hunk 7]
```diff
@@ -84,7 +87,9 @@ func (extr *Executor) UninstallKyma(installationData *config.InstallationData) e
 
 	removeLabelAndReturn := extr.removeLabelOnError(installationData.Context.Name, installationData.Context.Namespace)
 
-	err := extr.DeprovisionAzureResources(DefaultDeprovisionConfig(), installationData.Context)
+	err := extr.deprovisionAzureResourcesFn(extr, DefaultDeprovisionConfig(), installationData.Context)
+	//DeprovisionAzureResources(extr, DefaultDeprovisionConfig(), installationData.Context)
```
[the remark this answers is not in the document]
Comment (Tomasz-Smelcerz-SAP) [reply components/kyma-operator/pkg/kymaoperation/executor.go]: Done.

File: components/kyma-operator/pkg/kymahelm/wait.go (hunk 1 above)
Comment (jakkab) [reply components/kyma-operator/pkg/kymahelm/wait.go:92]: it's not resolved

Comment (jakkab) [reply components/kyma-operator/pkg/kymahelm/wait.go]: fixed and resolved

File: components/kyma-operator/pkg/kymahelm/kyma-helm-client.go [hunk 8]
```diff
@@ -112,122 +137,156 @@ func (hc *Client) IsReleaseDeletable(rname string) (bool, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	upgrade := action.NewUpgrade(cfg)
+	upgrade.Atomic = false
+	upgrade.CleanupOnFail = true
+	upgrade.Wait = true
+	upgrade.ReuseValues = true
+	upgrade.Recreate = true
```
[the remark this answers is not in the document]
Comment (jakkab) [reply components/kyma-operator/pkg/kymahelm/kyma-helm-client.go]: set to false

File: components/kyma-operator/pkg/kymaoperation/steps/creator.go [hunk 9]
```diff
@@ -25,35 +24,37 @@ func NewStepFactoryCreator(helmClient kymahelm.ClientInterface, sgls SourceGette
 	}
 }
 
+// TODO: Once migration to Helm3 is done, we should extend the map key to format: "namespace/name"
```
[the remark this answers is not in the document]
Comment (jakkab) [reply components/kyma-operator/pkg/kymaoperation/steps/creator.go:27]: @Tomasz-Smelcerz-SAP is it?

File: components/kyma-operator/README.md [hunk 10]
```diff
@@ -2,60 +2,65 @@
 
 ## Overview
 
-Kyma Operator is a tool for installing all Kyma components. The project is based on the Kubernetes operator pattern. It tracks changes of the Installation custom resource and installs, uninstalls, and updates Kyma accordingly.
+Kyma Operator is a tool for installing all Kyma components. The project is based on the Kubernetes operator pattern. It tracks changes of the Installation custom resource and installs, uninstalls, and updates Kyma accordingly. 
 
-## Prerequisites
+>**NOTE:** The Kyma Operator is designed to work as a [Singleton](https://en.wikipedia.org/wiki/Singleton_pattern), meaning that only one replica of the controller should be deployed, and only one Installation CR should be deployed
```
Comment (bszwarc) [inline components/kyma-operator/README.md:7]:
```suggestion
>**NOTE:** The Kyma Operator is designed to work as a [singleton](https://en.wikipedia.org/wiki/Singleton_pattern), meaning that you should deploly only one replica of the controller, and only one Installation CR.
```

File: components/kyma-operator/README.md [hunk 11]
```diff
@@ -2,60 +2,65 @@
...
 
-- Docker
+The controller extends the api-server with the following CustomResources:
```
Comment (bszwarc) [inline components/kyma-operator/README.md:9]:
```suggestion
The controller extends the api-server with the following custom resources:
```

File: components/kyma-operator/README.md [hunk 12]
```diff
@@ -2,60 +2,65 @@
...
+- `installation.installer.kyma-project.io/v1alpha1`
 
-## Development
+## Prerequisites
 
-Before each commit, use the [`Makefile`](./Makefile) script to test your changes:
-  ```bash
-  make verify
-  ```
+- recent version of Go language with support for modules (e.g: 1.12.6)
```
Comment (bszwarc) [inline components/kyma-operator/README.md:14]:
```suggestion
- recent version of Go with the support for modules, for example 1.12.6
```

File: components/kyma-operator/README.md [hunk 13]
```diff
@@ -2,60 +2,65 @@
...
+- make
+- kubectl
+- access to K8s environment: minikube or a remote K8s cluster
```
Comment (bszwarc) [inline components/kyma-operator/README.md:17]:
```suggestion
- access to Kubernetes environment: Minikube or a remote Kubernetes cluster
```

File: components/kyma-operator/README.md [hunk 14]
```diff
@@ -2,60 +2,65 @@
...
 
-## Rebuild Kubernetes client libraries
+## Details
 
-After introducing changes to the `apis/installer/v1alpha1` package, you have to regenerate client libraries by running:
-  ```
-  sh ./hack/update-codegen.sh
-  ```
+### Use command-line flags
```
Comment (bszwarc) [inline components/kyma-operator/README.md:21]:
```suggestion
### Use commandline flags
```

File: components/kyma-operator/README.md [hunk 15]
```diff
@@ -2,60 +2,65 @@
...
 
-## Build a Docker image
+| Name | Required | Description | Default value |
+|------|----------|-------------|-----------------|
```
Comment (bszwarc) [inline components/kyma-operator/README.md:24]:
```suggestion
|------|:----------:|-------------|-----------------|
```

File: components/kyma-operator/README.md [hunk 16]
```diff
@@ -2,60 +2,65 @@
...
+| **kubeconfig** | no | Path to a kubeconfig file | Taken from the `KUBECONFIG` env parameter|
+| **kymaDir** | no | Directory holding helm releases, within the image/pod | `/kyma` |
+| **backoffIntervals** | no | Number of seconds to wait before subsequent retries | `10,20,40,60,80` |
+| **overrideLogFile** | no | Log File to Print Installation overrides to. | `/dev/stdout` |
+| **overrideLogFormat** | no | Installation Override Log format (Accepted values: text or json) | `text` |
+| **helmMaxHistory**  | no | Max number of releases returned by Helm release history query | `10` |
+| **helmDriver** | no | Helm backed storage drivers. [Options and details](https://helm.sh/docs/helm/helm/#helm) | `secrets` |
+| **helmDebugMode** | no | Enable/Disable Helm client output in the Kyma Operator logs | `false` |
```
Comment (bszwarc) [inline components/kyma-operator/README.md:32]:
```suggestion
| **kubeconfig** | NO | Path to a `kubeconfig` file. | Taken from the **KUBECONFIG** environment parameter.|
| **kymaDir** | NO | Directory holding helm releases within the image or Pod. | `/kyma` |
| **backoffIntervals** | NO | Number of seconds to wait before subsequent retries. | `10,20,40,60,80` |
| **overrideLogFile** | NO | Log file to print installation overrides to. | `/dev/stdout` |
| **overrideLogFormat** | NO | Installation override log format. The accepted values are text or JSON. | `text` |
| **helmMaxHistory**  | NO | Maximum number of releases returned by the Helm release history query. | `10` |
| **helmDriver** | NO | Helm backed storage drivers. Read more about [options and details](https://helm.sh/docs/helm/helm/#helm). | `secrets` |
| **helmDebugMode** | NO | Enables/Disables Helm client output in the Kyma Operator logs. | `false` |
```

File: components/kyma-operator/README.md [hunk 17]
```diff
@@ -2,60 +2,65 @@
...
 
-Use the [`Makefile`](./Makefile) script to build a Docker image of the Kyma Operator:
-  ```
-  export DOCKER_PUSH_REPOSITORY={DOCKER_REPOSITORY}
-  make build-image
-  ```
+### Supported Environmental Variables
```
Comment (bszwarc) [inline components/kyma-operator/README.md:34]:
```suggestion
### Supported environment variables
```

File: components/kyma-operator/README.md [hunk 18]
```diff
@@ -2,60 +2,65 @@
...
 
-## Details
+| Name | Default | Description |
+| ---- | ------- | ----------- |
+| **INST_NAMESPACE** | `default` | Namespace in which the Installation CR is located |
+| **INST_RESOURCE** | `kyma-installation` | Name of the Installation custom resource |
+| **INST_FINALIZER** | `finalizer.installer.kyma-project.io` | Name of the finalizer |
+| **OVERRIDES_NAMESPACE** | `kyma-installer` | Namespace in which the Installer overrides are located |
```
Comment (bszwarc) [inline components/kyma-operator/README.md:41]:
```suggestion
| **INST_NAMESPACE** | `default` | Namespace in which the Installation CR is located. |
| **INST_RESOURCE** | `kyma-installation` | Name of the Installation custom resource. |
| **INST_FINALIZER** | `finalizer.installer.kyma-project.io` | Name of the finalizer. |
| **OVERRIDES_NAMESPACE** | `kyma-installer` | Namespace in which the Installer overrides are located. |
```

File: components/kyma-operator/README.md [hunk 19]
```diff
@@ -2,60 +2,65 @@
...
 
-### Upgrade Kyma
+## Custom Resource
```
Comment (bszwarc) [inline components/kyma-operator/README.md:43]:
```suggestion
## Installer custom resource
```

File: components/kyma-operator/README.md [hunk 20]
```diff
@@ -2,60 +2,65 @@
...
 
-The upgrade procedure relies heavily on Helm. When you trigger the upgrade, Helm performs `helm upgrade` on Helm releases that exist in the cluster and are defined in the Kyma version you're upgrading to. If a Helm release is defined in the Kyma version you're upgrading to but is not present in the cluster when you trigger the upgrade, Helm performs `helm install` and installs such a release.
+The `installation.installer.kyma-project.io` CustomResourceDefinition (CRD) is a detailed description of the kind of data and the format the Kyma Operator Controller listens for. To get the up-to-date CRD and show
```
Comment (bszwarc) [inline components/kyma-operator/README.md:45]: Is it the same custom resource as this one: https://kyma-project.io/docs/root/kyma#custom-resource-installation? 
Should it be `installation.installer.kyma-project.io` or `installations.installer.kyma-project.io`?

File: components/kyma-operator/README.md [hunk 21]
```diff
@@ -2,60 +2,65 @@
...
+## Additional information
 
-### Custom resource file
+### Upgrade Kyma
 
-The [Installation custom resource file](https://kyma-project.io/docs/root/kyma/#custom-resource-installation) provides the basic information for Kyma installation.
+The upgrade procedure relies heavily on Helm. When you trigger the upgrade, Helm performs `helm upgrade` on Helm releases that exist in the cluster and are defined in the Kyma version you're upgrading to. If a Helm release is defined in the Kyma version you're upgrading to but is not present in the cluster when you trigger the upgrade, Helm performs `helm install` and installs such a release.
 
-## Configuration
+When you trigger Kyma upgrade, the Kyma Operator lists the Helm releases installed in your current Kyma version. This list is compared against the list of Helm releases of the Kyma version you're upgrading to. The releases are matched by their names. The releases that match between versions are upgraded through `helm upgrade`. Releases that don't match are treated as new and installed through `helm install`.
```
Comment (bszwarc) [inline components/kyma-operator/README.md:62]:
```suggestion
When you trigger the Kyma upgrade, the Kyma Operator lists the Helm releases installed in your current Kyma version. This list is compared against the list of Helm releases of the Kyma version you're upgrading to, to match the releases by their names. The releases that match between versions are upgraded through `helm upgrade`. Releases that don't match are treated as new and installed through `helm install`.
```

File: components/kyma-operator/README.md [hunk 22]
```diff
@@ -2,60 +2,65 @@
...
-environmental variables:
+>**NOTE:** If you changed the name of a Helm release for a component, remove it before upgrading Kyma to prevent a situation where two Helm releases of the same component exist in the cluster.
 
-| Name                	| Default                               	| Description                                            	|
-|---------------------	|---------------------------------------	|--------------------------------------------------------	|
-| **INST_NAMESPACE**      	| `default`                             	| Namespace in which the Installation CR is located         	|
-| **INST_RESOURCE**       	| `kyma-installation`                   	| Name of the Installation custom resource                      	|
-| **INST_FINALIZER**      	| `finalizer.installer.kyma-project.io` 	| Name of the finalizer                                  	|
-| **OVERRIDES_NAMESPACE** 	| `kyma-installer`                      	| Namespace in which the Installer overrides are located 	|
+The Operator doesn't migrate Custom Resources to a new version when update is triggered. Custom Resource backward compatibility, or lack thereof, is determined at the component or Helm release level.
```
Comment (bszwarc) [inline components/kyma-operator/README.md:66]:
```suggestion
The Operator doesn't migrate custom resources to a new version when the upgrade is triggered. The custom resource backward compatibility, or lack thereof, is determined at the component or Helm release level.
```

File: components/kyma-operator/README.md [hunk 23]
```diff
@@ -2,60 +2,65 @@
...
-### Upgrade Kyma
+## Custom Resource
 
-The upgrade procedure relies heavily on Helm. When you trigger the upgrade, Helm performs `helm upgrade` on Helm releases that exist in the cluster and are defined in the Kyma version you're upgrading to. If a Helm release is defined in the Kyma version you're upgrading to but is not present in the cluster when you trigger the upgrade, Helm performs `helm install` and installs such a release.
+The `installation.installer.kyma-project.io` CustomResourceDefinition (CRD) is a detailed description of the kind of data and the format the Kyma Operator Controller listens for. To get the up-to-date CRD and show
+the output in the `yaml` format, run this command:
 
-When you trigger Kyma upgrade, the Kyma Operator lists the Helm releases installed in your current Kyma version. This list is compared against the list of Helm releases of the Kyma version you're upgrading to. The releases are matched by their names. The releases that match between versions are upgraded through `helm upgrade`. Releases that don't match are treated as new and installed through `helm install`.
+```bash
+kubectl get crd installation.installer.kyma-project.io -o yaml
```
Comment (bszwarc) [inline components/kyma-operator/README.md:49]: Again the question: `installation.installer.kyma-project.io` or `installations.installer.kyma-project.io`?

**Pull request merged**